### ISSUE TITLE
Add an option for Dection Models to return cost in eval mode as well (validation, test loss)

### ISSUE THREAD

**Issue by Delaunay** [pytorch/vision#1574]
We need to get the model cost in eval mode when we compute validation accuracy for hyper parameter search.

At the moment eval mode only returns the detection without the cost.
Which mean it is used as inference mode which is different than what people might want.

We can not use training mode to compute validation because of batchnorm and dropout.

The solution with the least code modification would be to add a new argument `inference_mode` in [forward][1] that would be true if we only need the detection and false if we want to cost.
and change `if self.training` to `if not inference_mode`.

[1]: https://github.com/pytorch/vision/blob/681c6c1115035a9dfd2f05ede1a00620532e2f01/torchvision/models/detection/generalized_rcnn.py#L31

[2]: https://github.com/pytorch/vision/blob/681c6c1115035a9dfd2f05ede1a00620532e2f01/torchvision/models/detection/generalized_rcnn.py#L59

**Comment by elangovana** [comment]
I would be interested in this feature too.

**Comment by fmassa** [comment]
Hi,

I think adding a new argument to the constructor might be a good way of handling this.
But given that this would be a breaking change, we will be doing it after the next release of torchvision, which will happen in January.

**Comment by ryzalk** [comment]
This would be a great feature. Looking forward to it :)

**Comment by cyyever** [comment]
I need this feature

**Comment by NicolasHug** [comment]
From offline discussions with @fmassa, we need the 3 following features:

- A way for eager and scripted mode to return the same type, i.e. a dict with losses and detections
- A way to return detections during training
- A way to return losses during eval

----

I think the following design might work:

- add new `return_tuples` param `{True, False (default)}`
  - possibly: add a `None` option that would be the default and that would warn users and tell them to explicitly set this parameter. But if we need to always support the legacy version anyway (by legacy I mean the current behaviour), this might be unnecessary.
  - When False, this corresponds to the legacy behaviour
  - When True, script and eager mode have identical behaviours (i.e. they return the same thing), as dictated by the next 2 parameters
- add new `compute_losses` param `{'always', 'training_only' (default)}`
   - if 'always':
        - `return_tuples` must be True
        - if targets aren't passed in eval, raise error (or at least warn)
- add new `compute_detections` param  `{'always', 'eval_only' (default)}`
  - if 'always', `return_tuples` must be True

It's 3 new parameters, but they're all backward-compatible so users only need to set them if they need/want to. And one advantage is that it's very explicit in what it does.

We could potentially avoid introducing the `compute_losses` parameter and basically return the losses iff `targets is not None`, but I fear this might be too implicit, and since we need to introduce `compute_detections` anyway, it's probably cleaner to do the same for losses.

Thoughts @fmassa ?

**Comment by NicolasHug** [comment]
From a discussion with @fmassa : there is hope that torchscript will eventually support returning union types. This would allow the scripted mode to return either the losses or the detections, instead of always returning a tuple, and would make eager and scripted versions compatible with a minimum amount of backward incompatibility. It may be preferable to wait until then, as the rest of the issues will be easier to address as well.

**Comment by bfialkoff** [comment]
I think Union types are now supported see [here](https://github.com/pytorch/pytorch/pull/64234)].

Also wouldn't it be simpler to just adjust the relevant `forward` methods to return losses if `targets` is provided. Then we could avoid discussions of `inference_mode` and just compute the losses if the targets are provided and return the detections always.

**Comment by khalidcawl** [comment]
Any updates on this? I looked around the code and couldn't find what to do if I want to get losses both during training and evaluation? Any help would be appreciated.

**Comment by claushofmann** [comment]
+ 1 for this issue. IMHO the most user-friendly solution (although I don't exactly know what would be the implications) would be to always return detections and losses if the user provides `(images, targets)` and to only return detections if the user provides `(images)`, disregarding whether the current mode is `training` or `eval`.

**Comment by deepwilson** [comment]
What's the update on this issue?

**Comment by bencevans** [comment]
I've just started trying to implement @NicolasHug's design specced out in  https://github.com/pytorch/vision/issues/1574#issuecomment-779360258.

https://github.com/pytorch/vision/compare/main...bencevans:vision:detection-output+loss

As @frgfm mentions in https://github.com/pytorch/vision/issues/1775 (duplicate issue), `RoiHeads` and `RegionProposalNetwork` will require changes, too, to force the loss to be returned. Is `force_loss: bool = False {default}` appropriate?

I'm pretty unfamiliar with the codebase (only as a user) so any pointers or collaboration would be awesome! I'd love to get this feature in as have been hit by it a few times now and end up hacking changes in mentioned from a related StackOverflow [thread](https://stackoverflow.com/questions/60339336/validation-loss-for-pytorch-faster-rcnn).

**Comment by Perry45** [comment]
Ok I came up with this idea for roi_heads.py:

`        if targets is not None:
            if self.training:
                proposals, matched_idxs, labels, regression_targets = self.select_training_samples(proposals, targets)
            else:
                proposalsV, matched_idxs, labels, regression_targets = self.select_training_samples(proposals, targets)
                box_featuresV = self.box_roi_pool(features, proposalsV, image_shapes)
                box_featuresV = self.box_head(box_featuresV)
                class_logitsV, box_regressionV = self.box_predictor(box_featuresV)
        else:
            labels = None
            regression_targets = None
            matched_idxs = None

        box_features = self.box_roi_pool(features, proposals, image_shapes)
        box_features = self.box_head(box_features)
        class_logits, box_regression = self.box_predictor(box_features)

        result: List[Dict[str, torch.Tensor]] = []
        losses = {}

        if labels is not None and regression_targets is not None:
            if self.training:
                loss_classifier, loss_box_reg = fastrcnn_loss(class_logits, box_regression, labels, regression_targets)
            else:
                loss_classifier, loss_box_reg = fastrcnn_loss(class_logitsV, box_regressionV, labels, regression_targets)
            losses = {"loss_classifier": loss_classifier, "loss_box_reg": loss_box_reg}

        if self.training:
            if labels is None:
                raise ValueError("labels cannot be None")
            if regression_targets is None:
                raise ValueError("regression_targets cannot be None")
        else:
            boxes, scores, labels = self.postprocess_detections(class_logits, box_regression, proposals, image_shapes)
            num_images = len(boxes)
            for i in range(num_images):
                result.append(
                    {
                        "boxes": boxes[i],
                        "labels": labels[i],
                        "scores": scores[i],
                    }
                )`

So if you are in training mode nothing changes. If you are in eval mode and no target is provided nothing changes. But if you are in eval mode and a target is passed, then I do the Pooling and stuff twice. Once for the loss with the data from select_training_samples and once with the original proposals to get the correct mAP so that it is consistent with the non-target call.
Is this a valid approach?

**Comment by NicolasHug** [comment]
@oliverdain reported in https://github.com/pytorch/vision/issues/7063 that there is a similar need for KeyPoint models

**Comment by WillieMaddox** [comment]
I've been working this problem too and wondering if it's even necessary that we call,
```
proposals, matched_idxs, labels, regression_targets = self.select_training_samples(proposals, targets)
```
when in `self.eval()` mode.  If so, which parts of `select_training_samples` are required for `self.eval()` mode?

For example, would adding gt_boxes to the proposals bias the val losses in some unexpected way?
```
proposals = self.add_gt_proposals(proposals, gt_boxes)
```
What about sampling a fixed proportion of positive-negative proposals.
```
sampled_inds = self.subsample(labels)
```
This cuts out a significant number of proposals which we may not want to do in eval mode.  It reminds me of Dropout.
Can anyone explain (or provide links to literature on) why these 2 methods are necessary?
